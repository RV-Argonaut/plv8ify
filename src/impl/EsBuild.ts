import { build } from 'esbuild'
import { BundleArgs, Bundler } from 'src/interfaces/Bundler'
import nodeExternals from 'webpack-node-externals'

class BundlerError extends Error {}

export class EsBuild implements Bundler {
  async bundle({ inputFile, globalName }: BundleArgs) {
    const esbuildResult = await build({
      entryPoints: [inputFile],
      globalName,
      platform: 'browser',
      external: [nodeExternals()],
      bundle: true,
      write: false,
      target: 'es2015',
    }).catch(() => new BundlerError('esbuild failed'))

    if (esbuildResult instanceof Error) {
      throw esbuildResult
    }

    const esbuildFile = esbuildResult.outputFiles.find((_) => true)
    const bundlesJs = esbuildFile.text
    return bundlesJs
  }
}
