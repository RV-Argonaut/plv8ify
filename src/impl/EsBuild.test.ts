import { describe, expect, it } from 'bun:test'

import { EsBuild } from './EsBuild'

describe('EsBuild tests', () => {
  it('getBundleJs - inline mode', async () => {
    const esbuild = new EsBuild()
    const js = await esbuild.bundle({
      inputFile: './src/test-fixtures/input.fixture.ts',
    })

    expect(js).toMatchSnapshot()
  })

  it('getBundleJs - bad syntax', async () => {
    const esbuild = new EsBuild()
    expect(
      esbuild.bundle({
        inputFile: './src/test-fixtures/bad.fixture.ts',
      })
    ).rejects.toMatchSnapshot()
  })

  it('getBundleJs - newline template', async () => {
    const esbuild = new EsBuild()
    const js = await esbuild.bundle({
      inputFile: './src/test-fixtures/newline-template.fixture.ts',
    })
    expect(js).toMatchSnapshot()
  })

  it('getBundleJs - newline string', async () => {
    const esbuild = new EsBuild()
    const js = await esbuild.bundle({
      inputFile: './src/test-fixtures/newline-string.fixture.ts',
    })
    expect(js).toMatchSnapshot()
  })
})
