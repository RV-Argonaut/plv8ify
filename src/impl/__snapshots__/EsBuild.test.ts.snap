// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`EsBuild tests getBundleJs - inline mode 1`] = `
"// src/test-fixtures/input.fixture.ts
function sayHello(toWhom) {
  const greet = \`hello \` + toWhom;
  return greet;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - newline template 1`] = `
"// src/test-fixtures/newline-template.fixture.ts
function sayHello() {
  const greetTemplate = \`hello 
 world\`;
  return greetTemplate;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - newline string 1`] = `
"// src/test-fixtures/newline-string.fixture.ts
function sayHello() {
  const greetString = "hello\nworld";
  return greetString;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - newline string 2`] = `
"// src/test-fixtures/newline-string.fixture.ts
function sayHello() {
  const greetString = "hello\nworld";
  return greetString;
}
export {
  sayHello
};
"
`;

exports[`EsBuild tests getBundleJs - bad syntax 1`] = `[BundlerError: esbuild failed]`;
